### PR TITLE
feat: `CodeHinter` with LSP visitor

### PR DESCRIPTION
**Pull request by LoZander**
This PR changes `CodeHinter` to use the LSP visitor instead of `Index`.

Related to #9508.

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 1]
```diff
@@ -16,349 +17,113 @@
...
-
-    case Expr.Discard(exp, _, _) =>
-      visitExp(exp)
-
-    case Expr.Match(matchExp, rules, _, _, _) =>
-      visitExp(matchExp) ++ rules.flatMap {
-        case MatchRule(_, guard, exp) => guard.toList.flatMap(visitExp) ::: visitExp(exp)
+    object HintConsumer extends Consumer {
+      override def consumeTraitSymUse(symUse: SymUse.TraitSymUse): Unit = {
+        hints = hints ++ considerTrait(root, symUse.sym, symUse.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:40]: Why interleave different concerns?

You could compute uses separate (in a helper function) and just use that

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:40]: Instead now your Consumers are producing hints. Seems like the opposite of separation of concerns.

Comment (LoZander) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:40]: Doing so would mean "throwing away" the type information we have at hand. When creating hints from these uses, we'd have to then pattern match to regain that information to find the def, enum or trait in question to gain it's annotation.

Ideas 
1. different collections for defs, traits, enums and def calls
2. one stack of `(Annotation, SourceLocation)` and treat all the same (check for each if it has an annotation). Separate stack for def call info `(Symbol, Type, SourceLocation)`

Comment (LoZander) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:40]: i agree that more separation would be good

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:40]: You need to collect uses right? Can you not write a visitors that collects those uses in separate maps without worrying that downstream they will be used for hints?

Comment (LoZander) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:40]: What I could do is collect every occurrence of every trait, enum, def use and def call, each in their own type of collection and then afterwards "consider" each. You mean something like that?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:40]: Yes

Comment (LoZander) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala]: How about now?

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 2]
```diff
@@ -15,350 +16,170 @@
...
-    case Body.Functional(_, exp, _) => visitExp(exp)
-    case Body.Guard(exp, _) => visitExp(exp)
+  private def checkParallel(ann: Annotations, loc: SourceLocation): List[CodeHint] = {
+    if (ann.isParallel) { CodeHint.Parallel(loc) :: Nil } else { Nil }
   }
 
   /**
     * Checks whether `sym` would benefit from `tpe` being pure.
     */
   private def checkEffect(sym: Symbol.DefnSym, tpe: Type, loc: SourceLocation)(implicit root: Root): List[CodeHint] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:164]: You should undo this change.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:164]: This PR is not about changing the meaning of the phase.

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 3]
```diff
@@ -15,350 +16,170 @@
...
-      visitExp(exp) ++ rules.flatMap {
-        case RestrictableChooseRule(_, body) => visitExp(body)
-      }
-
-    case Expr.Tag(_, exps, _, _, _) =>
-      visitExps(exps)
-
-    case Expr.RestrictableTag(_, exps, _, _, _) =>
-      visitExps(exps)
+  private case class CodeHintCandidates(traitOccurs: List[TraitSymUse], enumOccurs: List[(Symbol.EnumSym, SourceLocation)], defOccurs: List[DefSymUse], defCalls: List[(Symbol.DefnSym, List[Expr])])
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:110]: Lets sort the members alphabetically.

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 4]
```diff
@@ -15,350 +16,170 @@
...
+    * Returns a [[CodeHintCandidates]] for the Flix project.
+    *
+    * @param root The root AST node of the Flix project
+    * @return     A [[CodeHintCandidates]] for the Flix project.
+    */
+  private def getCandidates(implicit root: Root): CodeHintCandidates = {
+    var traitUses: List[TraitSymUse] = Nil
+    var defUses: List[DefSymUse] = Nil
+    var enumUses: List[(Symbol.EnumSym, SourceLocation)] = Nil
+    var defCalls: List[(Symbol.DefnSym, List[Expr])] = Nil
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:122]: sort fields alphabetically

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 5]
```diff
@@ -15,350 +16,170 @@
...
-    case Expr.TryCatch(exp, rules, _, _, _) =>
-      visitExp(exp) ++ rules.flatMap {
-        case CatchRule(_, _, exp) => visitExp(exp)
+    object UseConsumer extends Consumer {
+      override def consumeTraitSymUse(symUse: TraitSymUse): Unit = traitUses = symUse :: traitUses
+      override def consumeTraitConstraintHead(tcHead: TraitConstraint.Head): Unit = traitUses = TraitSymUse(tcHead.sym, tcHead.loc) :: traitUses
+      override def consumeDefSymUse(sym: DefSymUse): Unit = defUses = sym :: defUses
+      override def consumeExpr(exp: Expr): Unit = exp match {
+        case TypedAst.Expr.ApplyDef(symUse, exps, _, _, _, _) => defCalls = (symUse.sym, exps) :: defCalls
+        case _ => ()
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:130]: sort the above.

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 6]
```diff
@@ -15,350 +16,170 @@
...
 
-    case Expr.TryCatch(exp, rules, _, _, _) =>
-      visitExp(exp) ++ rules.flatMap {
-        case CatchRule(_, _, exp) => visitExp(exp)
+    object UseConsumer extends Consumer {
+      override def consumeTraitSymUse(symUse: TraitSymUse): Unit = traitUses = symUse :: traitUses
+      override def consumeTraitConstraintHead(tcHead: TraitConstraint.Head): Unit = traitUses = TraitSymUse(tcHead.sym, tcHead.loc) :: traitUses
+      override def consumeDefSymUse(sym: DefSymUse): Unit = defUses = sym :: defUses
+      override def consumeExpr(exp: Expr): Unit = exp match {
+        case TypedAst.Expr.ApplyDef(symUse, exps, _, _, _, _) => defCalls = (symUse.sym, exps) :: defCalls
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:129]: We don't use the term "call". We use "Apply" so call it applyDefs (this comment applies everywhere).

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 7]
```diff
@@ -15,350 +16,170 @@
...
-  private def visitEnum(enum0: TypedAst.Enum)(implicit index: Index): List[CodeHint] = {
-    val tagUses = enum0.cases.keys.flatMap(sym => index.usesOf(sym))
-    val enumUses = index.usesOf(enum0.sym)
-    val uses = enumUses ++ tagUses
-    val isDeprecated = enum0.ann.isDeprecated
-    val isExperimental = enum0.ann.isExperimental
-    val deprecated = if (isDeprecated) uses.map(CodeHint.Deprecated.apply) else Nil
-    val experimental = if (isExperimental) uses.map(CodeHint.Experimental.apply) else Nil
-    (deprecated ++ experimental).toList
+  private def considerDefCall(sym: Symbol.DefnSym, exps: List[Expr])(implicit root: Root): List[CodeHint] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:71]: These considerX should just be visitX as before

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 8]
```diff
@@ -15,350 +16,170 @@
...
   /**
     * Returns a collection of code quality hints for the given AST `root`.
     */
-  def run(root: TypedAst.Root, sources: Set[String])(implicit flix: Flix, index: Index): List[CodeHint] = {
-    val traitHints = root.traits.values.flatMap(visitTrait(_)(index)).toList
-    val defsHints = root.defs.values.flatMap(visitDef(_)(root, flix)).toList
-    val enumsHints = root.enums.values.flatMap(visitEnum(_)(index)).toList
-    (traitHints ++ defsHints ++ enumsHints).filter(include(_, sources))
+  def run(sources: Set[String])(implicit root: Root): List[CodeHint] = {
+    val cands = getCandidates
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:33]: why not something like: getOccurrences? or getUses?

Then the case class could be "Uses"

**Review by magnus-madsen — COMMENTED** [review]
Much better architecture. Just needs some cleanup and polish.

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala (hunk 2 above)
Comment (LoZander) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:164]: Yeah, okay. That's fair. If this indeed ought to be changed, it should be in a different PR. I get it :)

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala (hunk 8 above)
Comment (LoZander) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:33]: I did originally call it Uses, but I figured it didn't really reflect what it was that cleanly, since it also contains def application stuff, which are not really "uses" as such.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:33]: Occurrences then?

**Comment by LoZander** [comment]
Other than the one ongoing conversation, I've implemented the suggested changes :)

**Comment by magnus-madsen** [comment]
> Other than the one ongoing conversation, I've implemented the suggested changes :)

I suggest to call it Occurrences and then move on

**Comment by LoZander** [comment]
I think it's all done now :)

**Comment by LoZander** [comment]
Onto removing the index once and forall!

**Comment by LoZander** [comment]
(unless there's something else that should be addressed in this code)

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Merged 👍
